src/sallyport/mod.rs

```diff
@@ -174,7 +176,9 @@ impl Default for Block {
 impl Block {
     /// Returns the capacity of `Block.buf`
     pub const fn buf_capacity() -> usize {
-        512 * Page::size() - size_of::<Message>()
+        // FIXME: https://github.com/enarx/enarx-keepldr/issues/23
+        let page_num = if cfg!(test) { 1 } else { 512 };
+        page_num * Page::size() - size_of::<Message>()
```

Comment:
Why only for test builds?

Also, this is enormous, and we won't be able to construct the keep due to the 2MiB footprint assert in BootInfo::calculate.

512 * 4096 = 2097152 = 2 MiB (per CPU).

Which means that a single CPU configuration for the keep already depletes the quota enforced by BootInfo::calculate leaving no room for the shim or code.

This probably didn't come up in testing because the `kvm` backend mistakenly assumes that `size_of::<sallyport::Block>() == size_of::<Page>()`.

The following correction to the loader will uncover the assert failure.

```diff
diff --git a/src/backend/kvm/vm/builder.rs b/src/backend/kvm/vm/builder.rs
index 95e1001..8a40b97 100644
--- a/src/backend/kvm/vm/builder.rs
+++ b/src/backend/kvm/vm/builder.rs
@@ -113,7 +113,7 @@ impl<T: Hook> Builder<T> {
         code_size: Span<usize>,
     ) -> Result<BootInfo> {
         let setup_size = Page::size()
-            + (Page::size() * cfg.n_syscall_pages.get())
+            + (crate::sallyport::Block::buf_capacity() * cfg.n_syscall_pages.get())
             + size_of::<PageTable>()
             + size_of::<PageTable>();
```

I think the only other bugs with the loader WRT sallyport size will be `arch_specific_setup` since it writes a zeroed page to each sallyport block.

Reply:
Huh? I didn't change anything for the non-test case:

It was:
```diff
-        512 * Page::size() - size_of::<Message>()
```

Reply:
Yeah, but don't we want them to be the same size for test and non-test?

Reply:
We can't test big `Block` sizes on bare metal, because the `Block` is created on the stack.

Reply:
And, of course I will fix the whole thing with a follow-up patch. This is why the `FIXME` with a pointer to the issue is still in there.